File: backend/app/schemas/goal_chat.py
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


GoalIntent = Literal["target-date", "habit", "milestone"]
AccountabilityType = Literal["photo", "video", "voice", "check_in", "task_completion"]


class SelfProfilePrompt(BaseModel):
    prompt_id: str
    question: str


class GoalChatTask(BaseModel):
    name: str = Field(..., min_length=1)
    description: Optional[str] = None


class GoalChatSlotUpdates(BaseModel):
    intent: Optional[GoalIntent] = None
    success_definition: Optional[str] = None
    accountability_type: Optional[AccountabilityType] = None
    deadline: Optional[str] = None
    review_cycle: Optional[str] = None
    tasks: Optional[List[GoalChatTask]] = None
    user_summary: Optional[str] = None


class GoalChatProfileState(BaseModel):
    behavioral_summary: str
    self_profile_prompts: List[SelfProfilePrompt]
    answers: Dict[str, str]
    merged_summary: str


class GoalChatCreateSessionRequest(BaseModel):
    user_id: Optional[str] = None
    behavioral_summary: Optional[str] = None


class GoalChatCreateSessionResponse(BaseModel):
    session_id: str
    missing_slots: List[str]
    required_slots: List[str]
    profile: GoalChatProfileState


class GoalChatTurnRequest(BaseModel):
    message: str = Field(..., min_length=1)
    slot_updates: GoalChatSlotUpdates = Field(default_factory=GoalChatSlotUpdates)
    profile_answers: Dict[str, str] = Field(default_factory=dict)
    selected_chip: Optional[str] = None


class GoalChatTurnResponse(BaseModel):
    session_id: str
    missing_slots: List[str]
    captured_slots: Dict[str, Any]
    next_prompt: str
    is_ready_to_finalize: bool
    profile: GoalChatProfileState
    quick_reply_chips: List[str] = Field(default_factory=list)


class GoalChatFinalizeRequest(BaseModel):
    has_partner: bool


class GoalChatFinalizeResponse(BaseModel):
    session_id: str
    finalized: bool
    goal_plan: Optional[Dict[str, Any]] = None
    validation_errors: List[str] = Field(default_factory=list)


class GoalChatSummaryResponse(BaseModel):
    session_id: str
    ready_for_summary: bool
    summary: Dict[str, Any]


class GoalChatSummaryPatchRequest(BaseModel):
    summary: Dict[str, Any]


# ── Plan Generation Schemas ──


class PlanTask(BaseModel):
    name: str
    description: str = ""
    frequency: Literal["daily", "weekly", "biweekly", "monthly"] = "weekly"
    days: List[str] = Field(default_factory=list)
    duration_minutes: int = 30
    accountability_type: AccountabilityType = "task_completion"


class PlanMilestone(BaseModel):
    name: str
    description: str = ""
    target_week: int = 4
    progress_weight: int = 25
    tasks: List[PlanTask] = Field(default_factory=list)


class GeneratedPlan(BaseModel):
    title: str
    description: str = ""
    intent: GoalIntent = "habit"
    accountability_type: AccountabilityType = "task_completion"
    deadline: Optional[str] = None
    milestones: List[PlanMilestone] = Field(default_factory=list)


class GeneratePlanResponse(BaseModel):
    session_id: str
    plan: GeneratedPlan

File: src/lib/api/goal-chat.ts
import { apiClient } from "@/lib/api/client";

export type GoalChatTask = {
  name: string;
  description?: string;
  requires_partner_review?: boolean;
  review_sla?: string;
  escalation_policy?: string;
};

export type GoalChatProfileState = {
  behavioral_summary: string;
  self_profile_prompts: Array<{ prompt_id: string; question: string }>;
  answers: Record<string, string>;
  merged_summary: string;
};

export type GoalChatSessionResponse = {
  session_id: string;
  missing_slots: string[];
  required_slots: string[];
  profile: GoalChatProfileState;
};

export type GoalChatQuestionState = {
  missing_slots: string[];
  captured_slots: Record<string, any>;
  is_ready_to_finalize: boolean;
  profile: GoalChatProfileState;
};

export type GoalChatSummaryResponse = {
  session_id: string;
  ready_for_summary: boolean;
  summary: Record<string, any>;
};

export const createGoalChatSession = async (payload: { user_id?: string; behavioral_summary?: string }) =>
  apiClient.post<GoalChatSessionResponse>("/api/v1/goal-chat/sessions", payload);

export const getGoalChatSummary = async (sessionId: string) =>
  apiClient.post<GoalChatSummaryResponse>(`/api/v1/goal-chat/${sessionId}/summary`);

export const patchGoalChatSummary = async (sessionId: string, summary: Record<string, any>) =>
  apiClient.patch<GoalChatSummaryResponse>(`/api/v1/goal-chat/${sessionId}/summary`, { summary });

export const finalizeGoalChat = async (sessionId: string, hasPartner: boolean) =>
  apiClient.post<{ session_id: string; finalized: boolean; goal_plan?: Record<string, any>; validation_errors: string[] }>(
    `/api/v1/goal-chat/${sessionId}/finalize`,
    { has_partner: hasPartner },
  );

// ── Plan Generation Types ──

export type PlanTask = {
  name: string;
  description: string;
  frequency: "daily" | "weekly" | "biweekly" | "monthly";
  days: string[];
  duration_minutes: number;
  accountability_type: string;
};

export type PlanMilestone = {
  name: string;
  description: string;
  target_week: number;
  progress_weight: number;
  tasks: PlanTask[];
};

export type GeneratedPlan = {
  title: string;
  description: string;
  intent: string;
  accountability_type: string;
  deadline: string | null;
  milestones: PlanMilestone[];
};

export type GeneratePlanResponse = {
  session_id: string;
  plan: GeneratedPlan;
};

export const generateGoalPlan = async (sessionId: string) =>
  apiClient.post<GeneratePlanResponse>(`/api/v1/goal-chat/${sessionId}/generate-plan`);

export async function streamGoalChatTurn(
  sessionId: string,
  payload: { message: string; selected_chip?: string; profile_answers?: Record<string, string> },
  handlers: {
    onToken: (token: string) => void;
    onChips: (chips: string[]) => void;
    onQuestionState: (state: GoalChatQuestionState) => void;
    onReadyForSummary: (ready: boolean) => void;
    onDone?: () => void;
  },
  signal?: AbortSignal,
) {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL || "http://localhost:8000";

  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (typeof window !== "undefined") {
    const csrf = localStorage.getItem("csrf_token");
    if (csrf) headers["X-CSRF-Token"] = csrf;
  }

  const url = `${baseUrl}/api/v1/goal-chat/${sessionId}/turns/stream`;
  console.log("[GoalChat] streamGoalChatTurn: fetching", url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    credentials: "include",
    body: JSON.stringify({ message: payload.message, selected_chip: payload.selected_chip, profile_answers: payload.profile_answers || {} }),
    signal,
  });

  console.log("[GoalChat] streamGoalChatTurn: response status", response.status, "ok:", response.ok, "body:", !!response.body);

  if (!response.ok || !response.body) {
    throw new Error(`Streaming failed with status ${response.status}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) {
        console.log("[GoalChat] streamGoalChatTurn: stream done");
        break;
      }
      const chunk = decoder.decode(value, { stream: true });
      console.log("[GoalChat] streamGoalChatTurn: chunk received, length:", chunk.length);
      buffer += chunk;

      let boundary = buffer.indexOf("\n\n");
      while (boundary !== -1) {
        const rawEvent = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);
        boundary = buffer.indexOf("\n\n");

        const eventLine = rawEvent.split("\n").find((line) => line.startsWith("event:"));
        const dataLine = rawEvent.split("\n").find((line) => line.startsWith("data:"));
        if (!eventLine || !dataLine) continue;

        const eventName = eventLine.replace("event:", "").trim();
        const data = JSON.parse(dataLine.replace("data:", "").trim());

        console.log("[GoalChat] SSE event:", eventName);

        if (eventName === "token") handlers.onToken(data.text || "");
        if (eventName === "chips") handlers.onChips(data.chips || []);
        if (eventName === "question_state") handlers.onQuestionState(data);
        if (eventName === "ready_for_summary") handlers.onReadyForSummary(Boolean(data.ready));
        if (eventName === "done") handlers.onDone?.();
      }
    }
  } catch (err) {
    console.error("[GoalChat] streamGoalChatTurn: error", err);
    if (signal?.aborted) return; // user cancelled — not an error
    throw err;
  }
}
